Fix i386 xor encoder issue #1761

rhythmize opened the pull request:
This PR fixes issue #1761.

Encoder fiddler tries to encode avoid parameter if type is not bytes, but this raises exception if bytearray is passed.
Fixed by not encoding `avoid` parameter in `fiddler.py` if type is `bytearray` or `bytes`

rhythmize committed:
Fix i386 encoder issue #1761

Don't encode `avoid` parameter in `fiddler.py` if type is `bytearray`

Arusekk commented on pwnlib/util/fiddling.py:
```diff
@@ -370,7 +370,7 @@ def xor_pair(data, avoid = b'\x00\n'):
         data = packing.pack(data)
 
-    if not isinstance(avoid, bytes):
+    if not (isinstance(avoid, bytes) or isinstance(avoid, bytearray)):
```
More readable:
```suggestion
    if not isinstance(avoid, (bytes, bytearray)):
```

rhythmize replied:
Done https://github.com/rhythmize/pwntools/commit/9497263f3a9f3c74e4707ec4dbcc3505773ea58b

Arusekk reviewed (changes requested):
Great! I didn't know it takes so little to fix it.

rhythmize commented:
There's another issue if we directly call encode method, where `avoid` parameter is converted to `set` [here](https://github.com/Gallopsled/pwntools/blob/dev/pwnlib/encoders/encoder.py#L65) (I suppose to remove duplicates) and is passed to further functions [here](https://github.com/Gallopsled/pwntools/blob/dev/pwnlib/encoders/encoder.py#L83) as `set` instead of `bytes`. Will push that change as well.

rhythmize committed:
Pass `avoid` as `bytes` instead of `set` when `encoder.encode()` is called

Arusekk commented:
Is it not better to just support sets of bytes as well as bytes objects?

rhythmize commented:
In that case, I think we'll have to propogate the support for `set` type inside individual encoders across various architectures. From what I understand `encoder.encode()` is just a wrapper for actual encoders.
If individual encoders, already support `set` type along with `bytes`. Then we can use 
> if not isinstance(avoid, (bytes, bytearray, set)):

instead of 
> if not isinstance(avoid, (bytes, bytearray)):

If it doesn't break anything

rhythmize committed:
Requested change: Increase readability

heapcrash commented on pwnlib/encoders/encoder.py:
```diff
@@ -80,7 +80,7 @@ def encode(raw_bytes, avoid=None, expr=None, force=0, pcreg=''):
 
         try:
-            v = encoder(raw_bytes, avoid, pcreg)
+            v = encoder(raw_bytes, bytes(avoid), pcreg)
```
This should probably use `six.ensure_binary` instead of `bytes()`

heapcrash commented:
Does it make sense to convert `set()` to `bytes()` earlier in the pipeline?

rhythmize commented:
I tried it at earlier stage but then it crashes [here](https://github.com/Gallopsled/pwntools/blob/dev/pwnlib/encoders/encoder.py#L79).
Since `encoders.blacklist` is already declared as `set`, so for `&` operation to be valid `avoid` is required to be `set` and not `bytes`. Converting it to bytes beforehand leads to following error
> TypeError: unsupported operand type(s) for &: 'set' and 'bytes'

zachriggle commented on pwnlib/util/fiddling.py:
```diff
@@ -370,7 +370,7 @@ def xor_pair(data, avoid = b'\x00\n'):
+    if not isinstance(avoid, (bytes, bytearray)):
         avoid = avoid.encode('utf-8')
 
     avoid = bytearray(avoid)
```
If `avoid` is already a `bytearray` this will fail on the `encode`.  We should probably add a special case here too?

rhythmize replied:
Yeah, that's why we're not encoding if `avoid` is either `bytes` or `bytearray`.

heapcrash commented:
Looks like the CI failed on this, and the logs are gone now.  @rhythmize what's left to be done in this PR?

rhythmize commented:
From my side, I have made all the required changes to fix the original issue. And it's working fine for me. Can we re-trigger the CI pipeline to see what the error logs are so that I can debug it further?

Arusekk commented:
You can always merge latest dev or rebase on top of it, and the checks will run, too. I sometimes do it myself, but it requires a couple minutes for me to clone your repo and do all the job.

rhythmize commented:
Oh yeah sure. Thanks for the idea. I'll do that.

rhythmize commented:
Okay, looks like all the checks are passing now @Arusekk @heapcrash .

Arusekk reviewed (approved):
Some doctests would be great, but LGTM

rhythmize committed:
Update doctest in XOR encoder

Add my failing testcase to ensure the code is fixed

rhythmize commented:
@Arusekk I have updated the doctests in the XOR encoder since they were anyway failing for me due to import errors. And I have also added my test example to be sure that the fix has resolved the issue.

I was running it using `python -m doctest -v pwnlib/encoders/i386/xor.py`

Arusekk commented on pwnlib/encoders/i386/xor.py:
not too good, the tests should use `pwnlib` and not `pwn`, or relative access

Arusekk commented:
In order to run the tests, use `python3 -b -m sphinx -b doctest docs/source docs/build/doctest docs/source/encoders.rst`

You probably need to run `pip install -r docs/requirements.txt` before.

rhythmize commented:
Yup, thanks!
That works. I was running it from different workdir and config. So that's why it was failing for me.
Fixed now :)
